hazelcast/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceUnorderedP.java

```diff
@@ -95,6 +98,22 @@
     // TODO we can use more efficient structure: we only remove from the beginning and add to the end
     @SuppressWarnings("unchecked")
     private SortedMap<Long, Integer>[] watermarkCounts = new SortedMap[0];
+    /**
+     * Current inflight items.
+     * <p>
+     * Invariant: sum of all values in {@link #inFlightItems} is equal to
+     * {@link #asyncOpsCounter}
+     * <p>
+     * We keep items with the same key as single entry with counter. This may
+     * lead to starvation of watermarks in a sequence of events like:
```

Comment:
do we care? should be rather rare unless someone uses a source that emits the same item in constant interval which is later processed using `mapUsingServiceAsync`.

Reply:
I think your comment is wrong. I don't see any starvation. If the processing takes 10s but item is emitted every 5s, then `maxAsyncOps` will eventually be reached, no more items will be taken from the input, then input queue will fill up and the source will not be able to emit every 5s.

Also, WMs are emitted according to `watermarkCounts`, so in your scenario, the count of events before the watermark will get to 0, even if it's the same item all the time, and the WM will be forwarded. Can you reproduce this case?

Reply:
You are right, I mixed `inflightItems` and `watermarkCounts`